Implement Stein's algorithm for gcd

Emerentius opened the pull request:
The binary gcd algorithm offers significant speedups when the CPU has a builtin instruction for counting trailing zeros.

I benched the improvements with random numbers. Code is at  https://github.com/Emerentius/gcd_bench . These are only rough measurements, I used 50 (Edit: Now 5000) number pairs for each. The old gcd liked to crash for `i8` when I used more.

These are the improvements on my machine (Intel i5 2500k). All numbers are averaged.

```
u8
gcd:            29.65 ns / call
binary_gcd:     20.70 ns / call (  43.2% faster )

u16
gcd:            58.66 ns / call
binary_gcd:     42.50 ns / call (  38.0% faster )

u32
gcd:            96.71 ns / call
binary_gcd:     72.13 ns / call (  34.1% faster )

u64
gcd:           314.31 ns / call
binary_gcd:    148.77 ns / call ( 111.3% faster )

i8
gcd:            28.31 ns / call
binary_gcd:     17.33 ns / call (  63.3% faster )

i16
gcd:            62.20 ns / call
binary_gcd:     37.74 ns / call (  64.8% faster )

i32
gcd:           102.19 ns / call
binary_gcd:     65.84 ns / call (  55.2% faster )

i64
gcd:           388.04 ns / call
binary_gcd:    136.96 ns / call ( 183.3% faster )
```

As a bonus: `i8::min_value() % -1` and variants panic in the implementation of the euclidean algorithm. This implementation doesn't panic. Like in the euclidean implementation, the special case `gcd(i8::min_value(), i8::min_value()) == i8::min_value()` (negative output) still exists. Edit: In Debug mode it now panics.

I'm relying on signed integers to be in two's complement form.

Edit: 
* Updated bench numbers with 5000 random numbers for better statistics
* Added mention of debug mode panics on minimum value

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

Emerentius commented:
Noticed that `.abs()` panics on the minimum value in debug mode, allowing for a better guarantee of only positive return values, at least in debug mode. Added that to my implementation for signed integers.
Also added some tests
